mmdet/models/roi_heads/bbox_heads/multi_instance_bbox_head.py

```diff
@@ -0,0 +1,476 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule
+from mmengine.config import ConfigDict
+from mmengine.structures import InstanceData
+from torch import Tensor
+
+from mmdet.models.layers import multiclass_nms
+from mmdet.models.utils import empty_instances
+from mmdet.registry import MODELS
+from .bbox_head import BBoxHead
+
+
+def set_cpu_nms(pred_boxes, iou_threshold):
+    """Pure Python NMS baseline."""
+
+    def _overlap(det_boxes, base, others):
+        eps = 1e-8
+        x1_basement = det_boxes[base, 0]
+        y1_basement = det_boxes[base, 1]
+        x2_basement = det_boxes[base, 2]
+        y2_basement = det_boxes[base, 3]
+
+        x1_others = det_boxes[others, 0]
+        y1_others = det_boxes[others, 1]
+        x2_others = det_boxes[others, 2]
+        y2_others = det_boxes[others, 3]
+
+        areas_basement = (x2_basement - x1_basement) * (
+            y2_basement - y1_basement)
+        areas_others = (x2_others - x1_others) * (y2_others - y1_others)
+        xx1 = np.maximum(x1_basement, x1_others)
+        yy1 = np.maximum(y1_basement, y1_others)
+        xx2 = np.minimum(x2_basement, x2_others)
+        yy2 = np.minimum(y2_basement, y2_others)
+        w = np.maximum(0.0, xx2 - xx1)
+        h = np.maximum(0.0, yy2 - yy1)
+        inter = w * h
+        ovr = inter / (areas_basement + areas_others - inter + eps)
+        return ovr
+
+    scores = pred_boxes[:, 4]
+    order = np.argsort(-scores)
+    dets = pred_boxes[order]
+
+    numbers = dets[:, -1]
+    keep = np.ones(len(dets)) == 1
+    ruler = np.arange(len(dets))
+    while ruler.size > 0:
+        basement = ruler[0]
+        ruler = ruler[1:]
+        num = numbers[basement]
+        # calculate the body overlap
+        overlap = _overlap(dets[:, :4], basement, ruler)
+        indices = np.where(overlap > iou_threshold)[0]
+        loc = np.where(numbers[ruler][indices] == num)[0]
+        # the mask won't change in the step
+        mask = keep[ruler[indices][loc]]
+        keep[ruler[indices]] = False
+        keep[ruler[indices][loc][mask]] = True
+        ruler[~keep[ruler]] = -1
+        ruler = ruler[ruler > 0]
+    keep = keep[np.argsort(order)]
+    return keep
+
+
+def smooth_l1_loss(pred, target, beta: float):
```

Comment:
we have smooth l1 loss in the loss function

Reply:
I will replace this method with the method in mmdet. But I am aligning precision of CrowdDet. To find the source of the error, I prefer to use the original code in [CrowdDet](https://github.com/Purkialo/CrowdDet).